basenc: perform faster, streaming encoding

andrewliebenow opened the pull request:
Improve the performance, both in memory and time, of the encoding performed by the basenc (except in --z85 mode), base32, and base64 programs.

These programs now perform encoding in a buffered/streaming manner, so encoding is not constrained by the amount of available memory.

andrewliebenow commented:
# Setup

```
❯ dd if=/dev/urandom of=/dev/shm/one-random-gibibyte bs=1024 count=1048576

❯ du -k /dev/shm/one-random-gibibyte
1048576 /dev/shm/one-random-gibibyte

❯ cargo build --bin coreutils --features base64 --no-default-features --profile release
```

# No wrapping

## New implementation

Elapsed (wall clock) time (h:mm:ss or m:ss): **0:00.70**
Maximum resident set size (kbytes): **2716**

```
❯ /usr/bin/time --verbose -- ./target/release/coreutils base64 --wrap 0 -- /dev/shm/one-random-gibibyte 1>/dev/null
        Command being timed: "./target/release/coreutils base64 --wrap 0 -- /dev/shm/one-random-gibibyte"
        User time (seconds): 0.56
        System time (seconds): 0.13
        Percent of CPU this job got: 99%
        Elapsed (wall clock) time (h:mm:ss or m:ss): 0:00.70
        Average shared text size (kbytes): 0
        Average unshared data size (kbytes): 0
        Average stack size (kbytes): 0
        Average total size (kbytes): 0
        Maximum resident set size (kbytes): 2716
        Average resident set size (kbytes): 0
        Major (requiring I/O) page faults: 0
        Minor (reclaiming a frame) page faults: 119
        Voluntary context switches: 1
        Involuntary context switches: 1
        Swaps: 0
        File system inputs: 0
        File system outputs: 0
        Socket messages sent: 0
        Socket messages received: 0
        Signals delivered: 0
        Page size (bytes): 4096
        Exit status: 0
```

## Existing implementation

Elapsed (wall clock) time (h:mm:ss or m:ss): **0:01.99**
Maximum resident set size (kbytes): **2452180**

```
❯ /usr/bin/time --verbose -- coreutils base64 --wrap 0 -- /dev/shm/one-random-gibibyte 1>/dev/null
        Command being timed: "coreutils base64 --wrap 0 -- /dev/shm/one-random-gibibyte"
        User time (seconds): 0.61
        System time (seconds): 1.36
        Percent of CPU this job got: 99%
        Elapsed (wall clock) time (h:mm:ss or m:ss): 0:01.99
        Average shared text size (kbytes): 0
        Average unshared data size (kbytes): 0
        Average stack size (kbytes): 0
        Average total size (kbytes): 0
        Maximum resident set size (kbytes): 2452180
        Average resident set size (kbytes): 0
        Major (requiring I/O) page faults: 0
        Minor (reclaiming a frame) page faults: 611915
        Voluntary context switches: 1
        Involuntary context switches: 157
        Swaps: 0
        File system inputs: 0
        File system outputs: 0
        Socket messages sent: 0
        Socket messages received: 0
        Signals delivered: 0
        Page size (bytes): 4096
        Exit status: 0
```

## GNU Core Utilities's implementation

Elapsed (wall clock) time (h:mm:ss or m:ss): **0:00.72**
Maximum resident set size (kbytes): **1996**

```
❯ /usr/bin/time --verbose -- /usr/bin/base64 --wrap 0 -- /dev/shm/one-random-gibibyte 1>/dev/null
        Command being timed: "/usr/bin/base64 --wrap 0 -- /dev/shm/one-random-gibibyte"
        User time (seconds): 0.61
        System time (seconds): 0.10
        Percent of CPU this job got: 99%
        Elapsed (wall clock) time (h:mm:ss or m:ss): 0:00.72
        Average shared text size (kbytes): 0
        Average unshared data size (kbytes): 0
        Average stack size (kbytes): 0
        Average total size (kbytes): 0
        Maximum resident set size (kbytes): 1996
        Average resident set size (kbytes): 0
        Major (requiring I/O) page faults: 0
        Minor (reclaiming a frame) page faults: 103
        Voluntary context switches: 2
        Involuntary context switches: 14
        Swaps: 0
        File system inputs: 16
        File system outputs: 0
        Socket messages sent: 0
        Socket messages received: 0
        Signals delivered: 0
        Page size (bytes): 4096
        Exit status: 0
```

# Default wrapping (76 characters)

## New implementation

Elapsed (wall clock) time (h:mm:ss or m:ss): **0:03.40**
Maximum resident set size (kbytes): **2592**

```
❯ /usr/bin/time --verbose -- ./target/release/coreutils base64 -- /dev/shm/one-random-gibibyte 1>/dev/null
        Command being timed: "./target/release/coreutils base64 -- /dev/shm/one-random-gibibyte"
        User time (seconds): 3.22
        System time (seconds): 0.16
        Percent of CPU this job got: 99%
        Elapsed (wall clock) time (h:mm:ss or m:ss): 0:03.40
        Average shared text size (kbytes): 0
        Average unshared data size (kbytes): 0
        Average stack size (kbytes): 0
        Average total size (kbytes): 0
        Maximum resident set size (kbytes): 2592
        Average resident set size (kbytes): 0
        Major (requiring I/O) page faults: 0
        Minor (reclaiming a frame) page faults: 122
        Voluntary context switches: 1
        Involuntary context switches: 29
        Swaps: 0
        File system inputs: 0
        File system outputs: 0
        Socket messages sent: 0
        Socket messages received: 0
        Signals delivered: 0
        Page size (bytes): 4096
        Exit status: 0
```

## Existing implementation

Elapsed (wall clock) time (h:mm:ss or m:ss): **0:08.04**
Maximum resident set size (kbytes): **2452504**

```
❯ /usr/bin/time --verbose -- coreutils base64 -- /dev/shm/one-random-gibibyte 1>/dev/null
        Command being timed: "coreutils base64 -- /dev/shm/one-random-gibibyte"
        User time (seconds): 3.21
        System time (seconds): 4.78
        Percent of CPU this job got: 99%
        Elapsed (wall clock) time (h:mm:ss or m:ss): 0:08.04
        Average shared text size (kbytes): 0
        Average unshared data size (kbytes): 0
        Average stack size (kbytes): 0
        Average total size (kbytes): 0
        Maximum resident set size (kbytes): 2452504
        Average resident set size (kbytes): 0
        Major (requiring I/O) page faults: 0
        Minor (reclaiming a frame) page faults: 611918
        Voluntary context switches: 1
        Involuntary context switches: 491
        Swaps: 0
        File system inputs: 0
        File system outputs: 0
        Socket messages sent: 0
        Socket messages received: 0
        Signals delivered: 0
        Page size (bytes): 4096
        Exit status: 0
```

## GNU Core Utilities's implementation

Elapsed (wall clock) time (h:mm:ss or m:ss): **0:00.99**
Maximum resident set size (kbytes): **1912**

```
❯ /usr/bin/time --verbose -- /usr/bin/base64 -- /dev/shm/one-random-gibibyte 1>/dev/null
        Command being timed: "/usr/bin/base64 -- /dev/shm/one-random-gibibyte"
        User time (seconds): 0.79
        System time (seconds): 0.19
        Percent of CPU this job got: 99%
        Elapsed (wall clock) time (h:mm:ss or m:ss): 0:00.99
        Average shared text size (kbytes): 0
        Average unshared data size (kbytes): 0
        Average stack size (kbytes): 0
        Average total size (kbytes): 0
        Maximum resident set size (kbytes): 1912
        Average resident set size (kbytes): 0
        Major (requiring I/O) page faults: 0
        Minor (reclaiming a frame) page faults: 102
        Voluntary context switches: 1
        Involuntary context switches: 52
        Swaps: 0
        File system inputs: 0
        File system outputs: 0
        Socket messages sent: 0
        Socket messages received: 0
        Signals delivered: 0
        Page size (bytes): 4096
        Exit status: 0
```

sylvestre commented:
Please use hyperfine for benchmarking. Time isn't reliable enough for performances

andrewliebenow commented:
(Was having trouble getting `poop` to pipe output to /dev/null without using a shell script)

`benchmark.sh`:

```
#!/bin/sh

set \
	-C

set \
	-e

set \
	-f

set \
	-u

set \
	-x

${BASE_SIX_FOUR_BINARY:?} \
	--wrap \
	"${BASE_SIX_FOUR_WRAP_ARGUMENT:?}" \
	-- \
	/dev/shm/one-random-gibibyte \
	1>/dev/null
```

# No wrapping

## poop

### New implementation

```
❯ BASE_SIX_FOUR_BINARY='./target/release/coreutils base64' BASE_SIX_FOUR_WRAP_ARGUMENT='0' poop --duration 10000 './benchmark.sh'
Benchmark 1 (14 runs): ./benchmark.sh
  measurement          mean ± σ            min … max           outliers
  wall_time           724ms ± 8.12ms     711ms …  736ms          0 ( 0%)        
  peak_rss           3.48MB ± 99.1KB    3.33MB … 3.65MB          0 ( 0%)        
  cpu_cycles         2.19G  ± 12.2M     2.18G  … 2.22G           0 ( 0%)        
  instructions       9.55G  ±  167      9.55G  … 9.55G           0 ( 0%)        
  cache_references   35.7M  ± 1.44M     33.8M  … 38.1M           0 ( 0%)        
  cache_misses        328K  ± 24.6K      306K  …  385K           0 ( 0%)        
  branch_misses      2.33M  ± 3.70K     2.32M  … 2.33M           0 ( 0%)        
```

### Existing implementation

```
❯ BASE_SIX_FOUR_BINARY='coreutils base64' BASE_SIX_FOUR_WRAP_ARGUMENT='0' poop --duration 10000 './benchmark.sh' 
Benchmark 1 (5 runs): ./benchmark.sh
  measurement          mean ± σ            min … max           outliers
  wall_time          2.01s  ± 66.5ms    1.95s  … 2.13s           0 ( 0%)        
  peak_rss           2.51GB ±  108KB    2.51GB … 2.51GB          0 ( 0%)        
  cpu_cycles         2.38G  ± 27.6M     2.36G  … 2.42G           0 ( 0%)        
  instructions       9.40G  ±  370      9.40G  … 9.40G           0 ( 0%)        
  cache_references    116M  ±  320K      115M  …  116M           0 ( 0%)        
  cache_misses       1.44M  ± 46.4K     1.40M  … 1.51M           0 ( 0%)        
  branch_misses       372K  ±  294       372K  …  373K           0 ( 0%)        
```

## hyperfine

### New implementation

```
❯ BASE_SIX_FOUR_BINARY='./target/release/coreutils base64' BASE_SIX_FOUR_WRAP_ARGUMENT='0' hyperfine -- './benchmark.sh'         
Benchmark 1: ./benchmark.sh
  Time (mean ± σ):     714.1 ms ±  11.2 ms    [User: 535.9 ms, System: 172.8 ms]
  Range (min … max):   699.6 ms … 731.7 ms    10 runs
```

### Existing implementation

```
❯ BASE_SIX_FOUR_BINARY='coreutils base64' BASE_SIX_FOUR_WRAP_ARGUMENT='0' hyperfine -- './benchmark.sh'        
Benchmark 1: ./benchmark.sh
  Time (mean ± σ):      1.975 s ±  0.042 s    [User: 0.565 s, System: 1.396 s]
  Range (min … max):    1.929 s …  2.084 s    10 runs
 ```

andrewliebenow commented:
# No wrapping

```
❯ hyperfine ' ./target/release/coreutils base64 --wrap 0 -- /dev/shm/one-random-gibibyte 1>/dev/null ' ' coreutils base64 --wrap 0 -- /dev/shm/one-random-gibibyte 1>/dev/null '
Benchmark 1:  ./target/release/coreutils base64 --wrap 0 -- /dev/shm/one-random-gibibyte 1>/dev/null 
  Time (mean ± σ):     733.6 ms ±  21.8 ms    [User: 545.9 ms, System: 182.4 ms]
  Range (min … max):   710.7 ms … 772.3 ms    10 runs
 
Benchmark 2:  coreutils base64 --wrap 0 -- /dev/shm/one-random-gibibyte 1>/dev/null 
  Time (mean ± σ):      2.051 s ±  0.022 s    [User: 0.565 s, System: 1.469 s]
  Range (min … max):    2.013 s …  2.082 s    10 runs
 
Summary
   ./target/release/coreutils base64 --wrap 0 -- /dev/shm/one-random-gibibyte 1>/dev/null  ran
    2.80 ± 0.09 times faster than  coreutils base64 --wrap 0 -- /dev/shm/one-random-gibibyte 1>/dev/null 
```

# Default wrapping (76 characters)

```
❯ hyperfine ' ./target/release/coreutils base64 -- /dev/shm/one-random-gibibyte 1>/dev/null ' ' coreutils base64 -- /dev/shm/one-random-gibibyte 1>/dev/null '           
Benchmark 1:  ./target/release/coreutils base64 -- /dev/shm/one-random-gibibyte 1>/dev/null 
  Time (mean ± σ):      3.288 s ±  0.157 s    [User: 3.093 s, System: 0.177 s]
  Range (min … max):    3.023 s …  3.461 s    10 runs
 
Benchmark 2:  coreutils base64 -- /dev/shm/one-random-gibibyte 1>/dev/null 
  Time (mean ± σ):      8.115 s ±  0.067 s    [User: 3.234 s, System: 4.831 s]
  Range (min … max):    8.006 s …  8.220 s    10 runs
 
Summary
   ./target/release/coreutils base64 -- /dev/shm/one-random-gibibyte 1>/dev/null  ran
    2.47 ± 0.12 times faster than  coreutils base64 -- /dev/shm/one-random-gibibyte 1>/dev/null 
```

andrewliebenow commented:
## Decoding performance data

### Setup

```
❯ cargo \
        build \
        --bin \
        coreutils \
        --features \
        base32 \
        --features \
        base64 \
        --features \
        basenc \
        --no-default-features \
        --profile release

❯ dd if=/dev/urandom of=/dev/stdout bs=1024 count=1048576 | /usr/bin/base64 --wrap 0 -- - 1>/dev/shm/one-random-gibibyte-base-six-four-encoded

❯ du -k -- /dev/shm/one-random-gibibyte-base-six-four-encoded
1398104 /dev/shm/one-random-gibibyte-base-six-four-encoded
```

### Hyperfine

```
❯ hyperfine \
        -- \
        ' ./target/release/coreutils base64 --decode -- /dev/shm/one-random-gibibyte-base-six-four-encoded 1>/dev/null ' \
        ' coreutils base64 --decode -- /dev/shm/one-random-gibibyte-base-six-four-encoded 1>/dev/null ' \
        ' /usr/bin/base64 --decode -- /dev/shm/one-random-gibibyte-base-six-four-encoded 1>/dev/null '

Benchmark 1:  ./target/release/coreutils base64 --decode -- /dev/shm/one-random-gibibyte-base-six-four-encoded 1>/dev/null 
  Time (mean ± σ):     906.3 ms ±  10.7 ms    [User: 631.3 ms, System: 269.8 ms]
  Range (min … max):   893.4 ms … 920.2 ms    10 runs
 
Benchmark 2:  coreutils base64 --decode -- /dev/shm/one-random-gibibyte-base-six-four-encoded 1>/dev/null 
  Time (mean ± σ):      2.815 s ±  0.033 s    [User: 1.274 s, System: 1.524 s]
  Range (min … max):    2.775 s …  2.871 s    10 runs
 
Benchmark 3:  /usr/bin/base64 --decode -- /dev/shm/one-random-gibibyte-base-six-four-encoded 1>/dev/null 
  Time (mean ± σ):      1.652 s ±  0.036 s    [User: 1.338 s, System: 0.305 s]
  Range (min … max):    1.611 s …  1.733 s    10 runs
 
Summary
   ./target/release/coreutils base64 --decode -- /dev/shm/one-random-gibibyte-base-six-four-encoded 1>/dev/null  ran
    1.82 ± 0.04 times faster than  /usr/bin/base64 --decode -- /dev/shm/one-random-gibibyte-base-six-four-encoded 1>/dev/null 
    3.11 ± 0.05 times faster than  coreutils base64 --decode -- /dev/shm/one-random-gibibyte-base-six-four-encoded 1>/dev/null 
```

### poop

```
❯ poop \
        ' ./target/release/coreutils base64 --decode -- /dev/shm/one-random-gibibyte-base-six-four-encoded ' \
        ' coreutils base64 --decode -- /dev/shm/one-random-gibibyte-base-six-four-encoded ' \
        ' /usr/bin/base64 --decode -- /dev/shm/one-random-gibibyte-base-six-four-encoded '

Benchmark 1 (6 runs): ./target/release/coreutils base64 --decode -- /dev/shm/one-random-gibibyte-base-six-four-encoded
  measurement          mean ± σ            min … max           outliers         delta
  wall_time           905ms ± 13.7ms     893ms …  920ms          0 ( 0%)        0%
  peak_rss           2.73MB ± 45.7KB    2.68MB … 2.79MB          0 ( 0%)        0%
  cpu_cycles         2.49G  ± 3.69M     2.49G  … 2.50G           0 ( 0%)        0%
  instructions       11.1G  ±  196      11.1G  … 11.1G           0 ( 0%)        0%
  cache_references   20.9M  ±  896K     20.1M  … 22.1M           0 ( 0%)        0%
  cache_misses        519K  ± 23.9K      494K  …  556K           0 ( 0%)        0%
  branch_misses      2.88M  ± 2.87K     2.87M  … 2.88M           0 ( 0%)        0%
Benchmark 2 (3 runs): coreutils base64 --decode -- /dev/shm/one-random-gibibyte-base-six-four-encoded
  measurement          mean ± σ            min … max           outliers         delta
  wall_time          2.81s  ± 30.0ms    2.78s  … 2.84s           0 ( 0%)        💩+211.0% ±  3.7%
  peak_rss           2.51GB ±  144KB    2.51GB … 2.51GB          0 ( 0%)        💩+92020.0% ±  5.3%
  cpu_cycles         5.29G  ± 30.4M     5.26G  … 5.32G           0 ( 0%)        💩+112.2% ±  1.1%
  instructions       22.2G  ± 37.1      22.2G  … 22.2G           0 ( 0%)        💩+ 99.5% ±  0.0%
  cache_references    118M  ±  513K      118M  …  119M           0 ( 0%)        💩+465.3% ±  6.4%
  cache_misses        946K  ± 1.87K      945K  …  948K           0 ( 0%)        💩+ 82.4% ±  6.5%
  branch_misses       272K  ±  812       271K  …  273K           0 ( 0%)        ⚡- 90.5% ±  0.1%
Benchmark 3 (4 runs): /usr/bin/base64 --decode -- /dev/shm/one-random-gibibyte-base-six-four-encoded
  measurement          mean ± σ            min … max           outliers         delta
  wall_time          1.63s  ± 40.5ms    1.60s  … 1.69s           0 ( 0%)        💩+ 79.7% ±  4.5%
  peak_rss           2.00MB ± 45.5KB    1.97MB … 2.06MB          0 ( 0%)        ⚡- 26.7% ±  2.5%
  cpu_cycles         5.39G  ±  151M     5.29G  … 5.61G           0 ( 0%)        💩+116.2% ±  5.5%
  instructions       25.3G  ±  108      25.3G  … 25.3G           0 ( 0%)        💩+127.8% ±  0.0%
  cache_references   7.61M  ±  629K     7.10M  … 8.48M           0 ( 0%)        ⚡- 63.7% ±  5.7%
  cache_misses        518K  ± 2.22K      515K  …  520K           1 (25%)          -  0.1% ±  5.4%
  branch_misses      4.08M  ± 15.1K     4.06M  … 4.10M           0 ( 0%)        💩+ 41.8% ±  0.5%
```

andrewliebenow commented:
Last one:

```
❯ poop \
        ' ./target/release/coreutils base64 --decode --ignore-garbage -- /dev/shm/one-random-gibibyte-base-six-four-encoded ' \
        ' coreutils base64 --decode --ignore-garbage -- /dev/shm/one-random-gibibyte-base-six-four-encoded ' \
        ' /usr/bin/base64 --decode --ignore-garbage -- /dev/shm/one-random-gibibyte-base-six-four-encoded '

Benchmark 1 (4 runs): ./target/release/coreutils base64 --decode --ignore-garbage -- /dev/shm/one-random-gibibyte-base-six-four-encoded
  measurement          mean ± σ            min … max           outliers         delta
  wall_time          1.47s  ± 12.7ms    1.46s  … 1.49s           0 ( 0%)        0%
  peak_rss           2.73MB ± 41.1KB    2.67MB … 2.76MB          1 (25%)        0%
  cpu_cycles         4.88G  ± 35.6M     4.86G  … 4.93G           0 ( 0%)        0%
  instructions       21.1G  ±  272      21.1G  … 21.1G           0 ( 0%)        0%
  cache_references   25.9M  ± 1.17M     24.7M  … 27.5M           0 ( 0%)        0%
  cache_misses        484K  ± 19.0K      457K  …  499K           1 (25%)        0%
  branch_misses      3.05M  ± 1.69K     3.05M  … 3.05M           0 ( 0%)        0%
Benchmark 2 (3 runs): coreutils base64 --decode --ignore-garbage -- /dev/shm/one-random-gibibyte-base-six-four-encoded
  measurement          mean ± σ            min … max           outliers         delta
  wall_time          31.6s  ± 60.1ms    31.5s  … 31.6s           0 ( 0%)        💩+2045.4% ±  5.2%
  peak_rss           2.51GB ± 35.5KB    2.51GB … 2.51GB          0 ( 0%)        💩+91805.4% ±  2.8%
  cpu_cycles          125G  ± 82.8M      125G  …  125G           0 ( 0%)        💩+2455.1% ±  2.4%
  instructions        174G  ±  492       174G  …  174G           0 ( 0%)        💩+723.0% ±  0.0%
  cache_references    121M  ±  433K      120M  …  121M           0 ( 0%)        💩+365.6% ±  7.2%
  cache_misses       1.32M  ± 52.3K     1.26M  … 1.36M           0 ( 0%)        💩+172.3% ± 14.7%
  branch_misses      2.35G  ± 13.8M     2.34G  … 2.36G           0 ( 0%)        💩+76749.9% ± 559.8%
Benchmark 3 (3 runs): /usr/bin/base64 --decode --ignore-garbage -- /dev/shm/one-random-gibibyte-base-six-four-encoded
  measurement          mean ± σ            min … max           outliers         delta
  wall_time          2.30s  ± 58.5ms    2.24s  … 2.36s           0 ( 0%)        💩+ 56.6% ±  5.1%
  peak_rss           1.99MB ± 52.2KB    1.96MB … 2.05MB          0 ( 0%)        ⚡- 27.1% ±  3.3%
  cpu_cycles         8.11G  ±  235M     7.86G  … 8.32G           0 ( 0%)        💩+ 66.2% ±  6.1%
  instructions       36.8G  ±  128      36.8G  … 36.8G           0 ( 0%)        💩+ 74.1% ±  0.0%
  cache_references   10.5M  ± 1.26M     9.63M  … 12.0M           0 ( 0%)        ⚡- 59.3% ±  9.2%
  cache_misses        629K  ±  104K      532K  …  738K           0 ( 0%)        💩+ 29.8% ± 27.2%
  branch_misses      4.46M  ± 25.9K     4.45M  … 4.49M           0 ( 0%)        💩+ 46.2% ±  1.1%
```

andrewliebenow commented:
Encoding with line wrapping enabled was the last remaining scenario I'm aware of in which GNU Core Utilities was still faster, and now that's been fixed:

### hyperfine

```
❯ hyperfine \
        ' ./target/release/coreutils base64 --wrap 76 -- /dev/shm/one-random-gibibyte 1>/dev/null ' \
        ' coreutils base64 --wrap 76 -- /dev/shm/one-random-gibibyte 1>/dev/null ' \
        ' /usr/bin/base64 --wrap 76 -- /dev/shm/one-random-gibibyte 1>/dev/null '
Benchmark 1:  ./target/release/coreutils base64 --wrap 76 -- /dev/shm/one-random-gibibyte 1>/dev/null 
  Time (mean ± σ):     774.4 ms ±  18.6 ms    [User: 603.9 ms, System: 165.0 ms]
  Range (min … max):   761.5 ms … 821.8 ms    10 runs
 
Benchmark 2:  coreutils base64 --wrap 76 -- /dev/shm/one-random-gibibyte 1>/dev/null 
  Time (mean ± σ):      8.064 s ±  0.146 s    [User: 3.241 s, System: 4.766 s]
  Range (min … max):    7.875 s …  8.266 s    10 runs
 
Benchmark 3:  /usr/bin/base64 --wrap 76 -- /dev/shm/one-random-gibibyte 1>/dev/null 
  Time (mean ± σ):      1.038 s ±  0.022 s    [User: 0.840 s, System: 0.188 s]
  Range (min … max):    1.017 s …  1.091 s    10 runs
 
Summary
   ./target/release/coreutils base64 --wrap 76 -- /dev/shm/one-random-gibibyte 1>/dev/null  ran
    1.34 ± 0.04 times faster than  /usr/bin/base64 --wrap 76 -- /dev/shm/one-random-gibibyte 1>/dev/null 
   10.41 ± 0.31 times faster than  coreutils base64 --wrap 76 -- /dev/shm/one-random-gibibyte 1>/dev/null 
```

### poop

```
❯ poop \
    ' coreutils base64 --wrap 76 -- /dev/shm/one-random-gibibyte ' \
    ' /usr/bin/base64 --wrap 76 -- /dev/shm/one-random-gibibyte ' \
    ' ./target/release/coreutils base64 --wrap 76 -- /dev/shm/one-random-gibibyte '

Benchmark 1 (3 runs): coreutils base64 --wrap 76 -- /dev/shm/one-random-gibibyte
  measurement          mean ± σ            min … max           outliers         delta
  wall_time          7.99s  ± 72.5ms    7.93s  … 8.07s           0 ( 0%)        0%
  peak_rss           2.51GB ±  207KB    2.51GB … 2.51GB          0 ( 0%)        0%
  cpu_cycles         10.4G  ± 72.8M     10.3G  … 10.4G           0 ( 0%)        0%
  instructions       20.8G  ±  306      20.8G  … 20.8G           0 ( 0%)        0%
  cache_references    120M  ±  496K      120M  …  121M           0 ( 0%)        0%
  cache_misses       1.96M  ± 91.7K     1.86M  … 2.02M           0 ( 0%)        0%
  branch_misses       143M  ±  117K      142M  …  143M           0 ( 0%)        0%
Benchmark 2 (5 runs): /usr/bin/base64 --wrap 76 -- /dev/shm/one-random-gibibyte
  measurement          mean ± σ            min … max           outliers         delta
  wall_time          1.03s  ± 32.4ms    1.00s  … 1.08s           0 ( 0%)        ⚡- 87.1% ±  1.1%
  peak_rss           2.05MB ± 35.9KB    2.00MB … 2.08MB          0 ( 0%)        ⚡- 99.9% ±  0.0%
  cpu_cycles         3.35G  ± 98.7M     3.29G  … 3.52G           0 ( 0%)        ⚡- 67.8% ±  1.6%
  instructions       14.5G  ± 70.1      14.5G  … 14.5G           0 ( 0%)        ⚡- 30.3% ±  0.0%
  cache_references   72.7M  ± 3.23M     69.7M  … 77.8M           0 ( 0%)        ⚡- 39.5% ±  3.9%
  cache_misses        524K  ±  151K      339K  …  720K           0 ( 0%)        ⚡- 73.3% ± 12.2%
  branch_misses      3.64M  ± 7.71K     3.63M  … 3.65M           0 ( 0%)        ⚡- 97.4% ±  0.1%
Benchmark 3 (7 runs): ./target/release/coreutils base64 --wrap 76 -- /dev/shm/one-random-gibibyte
  measurement          mean ± σ            min … max           outliers         delta
  wall_time           778ms ± 14.8ms     763ms …  807ms          0 ( 0%)        ⚡- 90.3% ±  0.8%
  peak_rss           2.84MB ± 67.6KB    2.70MB … 2.90MB          0 ( 0%)        ⚡- 99.9% ±  0.0%
  cpu_cycles         2.46G  ± 31.4M     2.44G  … 2.53G           0 ( 0%)        ⚡- 76.3% ±  0.7%
  instructions       9.60G  ±  289      9.60G  … 9.60G           0 ( 0%)        ⚡- 53.7% ±  0.0%
  cache_references   93.3M  ± 2.01M     91.3M  … 97.4M           0 ( 0%)        ⚡- 22.4% ±  2.3%
  cache_misses        397K  ± 43.8K      339K  …  476K           0 ( 0%)        ⚡- 79.8% ±  4.8%
  branch_misses      2.58M  ± 11.1K     2.56M  … 2.60M           0 ( 0%)        ⚡- 98.2% ±  0.1%
```

BenWiederhake commented:
Not directly related, but I'm mentioning it for completeness: #6008

BenWiederhake commented on src/uucore/src/lib/features/encoding.rs:
```diff
@@ -42,138 +30,115 @@ pub enum Format {
-            }
-        }
-    })
+pub struct ZEightFiveWrapper {}
```
Is there a reason why this isn't just called `Z85Wrapper`, as one would expect for z85 encoding?

andrewliebenow replied:
Will change

BenWiederhake commented on src/uucore/src/lib/features/encoding.rs:
```diff
@@ -42,138 +30,115 @@ pub enum Format {
-    }
+        // According to the spec we should not accept inputs whose len is not a multiple of 4.
+        // However, the z85 crate implements a padded encoding and accepts such inputs. We have to manually check for them.
+        if input.len() % 4_usize != 0_usize {
```
Again, why specify the integer types explicitly? Please write `4` instead of `4_usize`, or document the reason why this is necessary.

andrewliebenow replied:
Will change

BenWiederhake commented on src/uu/base32/src/base_common.rs:
```diff
@@ -134,52 +141,615 @@ pub fn base_app(about: &'static str, usage: &str) -> Command {
+            write_to_stdout(&mut decoded_buffer, &mut stdout_lock)?;
+        }
+
+        Ok(())
```
That's a lot of new complex code in `base_common.rs`. Could you add a few unit-tests, just to sanity-check everything? I know that it (seems to) work just fine, and skimming the code I didn't spot any obvious flaws.

- I'm sure there were a few edge cases where you had to take extra precaution. Those are ideal things to test for, because we want to make sure that now only is the *current* code correct, but also gives us assurance that future code *remains* correct.
- Also, Just basic functionality. This makes it easier to spot any issues at the right level, not only during the integration test of the base64 program (for example).

andrewliebenow replied:
What do you think of using fuzz or property testing? There was one attempted optimization I had to roll back because of a bug that only manifested on large inputs. The problem is that, AFAIK, the project is not using fuzz or property testing anywhere else, so it might be a bit out of place here.

BenWiederhake commented on src/uucore/src/lib/features/encoding.rs:
```diff
@@ -42,138 +30,115 @@ pub enum Format {
-        writeln!(writer, "{}", &res[start..end])?;
-        start = end;
+        let decode_result = match z85::decode(input) {
+            Ok(ve) => ve,
```
Codecov says that this (and many other) lines weren't tested. This is surprising. Is this a false negative, or is there really no test for many lines of code in this PR?

andrewliebenow replied:
I'm not exactly sure how this logic would be unit tested, or how beneficial that would be compared to other approaches. I do agree that it would be nice to have more assurance that these new implementations are correct.

The tests in `test_base32.rs`, `test_base64.rs`, and `test_basenc.rs` are covering basic uses with small inputs, but because of the new buffering logic, I think fuzz or property testing, which could generate large random inputs, would be a good way to make sure that there are no subtle edge cases.

BenWiederhake commented on tests/by-util/test_basenc.rs:
```diff
@@ -26,7 +31,9 @@ fn test_invalid_input() {
-        "basenc: error: invalid input\n"
+        // TODO
+        // Other implementations do not show " (os error 21)"
+        "basenc: read error: Is a directory (os error 21)\n"
```
"os error 21" probably refers to the UNIX errno, and shouldn't appear in a platform-independent test. I suggest changing the error message and removing the "os error" entirely. After all, GNU base64 doesn't print it either, and we try to be as close to it as reasonably possible:

```console
$ base64 /tmp 
base64: read error: Is a directory
[$? = 1]
```

BenWiederhake reviewed (changes requested):
The code looks good, and your performance comparison looks awesome, thank you!

I have some suggestions about some names and style decisions that I feel are way too unusual and surprising, and would like to ask you to change them.

Also, Codecov complains about lines being untested, and there seem to be no unit test for the hundreds of new lines. Please fix that.

This PR improves `base*` substantially, and I'm looking forward to seeing it merged!

andrewliebenow commented on src/uu/base32/src/base_common.rs:
```diff
@@ -134,52 +141,615 @@ pub fn base_app(about: &'static str, usage: &str) -> Command {
+
+        Ok(())
+    } else {
+        let (encoding, un_padded_multiple) = match format {
```
I was trying to cheat CSpell here. I'll add a spelling ignore comment instead.

andrewliebenow commented:
> Not directly related, but I'm mentioning it for completeness: #6008

I'd prefer to handle that in a separate PR to limit this PR's scope, since AFAIK, this PR does not change uutils' behavior re: invalid inputs. It does seem like these programs would be more useful if they displayed partial output, since often users just want to have some idea of what the data they have is, even if it's not totally valid.

andrewliebenow commented:
I have added property testing of the core decoding/encoding logic to the PR. Doing so helped catch two bugs: one that I introduced in `cc9d92cecb7446942579196e6c8d46c2f9824149` while extracting some logic into a helper function (I accidentally transposed two arguments). This bug was not caught by the existing unit tests, because it only manifested on large-ish inputs.

The other bug was related to my misunderstanding of input length validation for Z85.

Anyway, this definitely sold me on property testing. I have tuned the parameters so that the property tests for `base64`/`base32`/`basenc` complete within a minute (on my machine, at least).

sylvestre commented:
Could you please add your methodology into src/uu/base64/BENCHMARKING.md like we are doing for other programs?

thanks

andrewliebenow commented:
Benchmarking instructions, and some benchmark results, added:

https://github.com/uutils/coreutils/blob/7672dd926b1da4af4bf21620faee9e6b99813b2c/src/uu/basenc/BENCHMARKING.md

sylvestre commented:
great doc, it is a nice read :)

sylvestre commented on src/uu/basenc/BENCHMARKING.md:
```diff
@@ -0,0 +1,177 @@
+  Range (min … max):   683.1 ms … 713.5 ms    10 runs
+
+Relative speed comparison
+        1.39 ±  0.02  /usr/bin/basenc --base64 -- /dev/shm/one-random-gibibyte 1>/dev/null
```
I can't reproduce this:

```
Relative speed comparison
        1.00          /usr/bin/basenc --base16 --wrap 0 -- /dev/shm/one-random-gibibyte
        1.22 ±  0.35  ./target/release/coreutils basenc --base16 --wrap 0 -- /dev/shm/one-random-gibibyte
        3.66 ±  1.22  ./coreutils basenc --base16 --wrap 0 -- /dev/shm/one-random-gibibyte
```

./coreutils is the version without the patch

sylvestre replied:
which is fine btw :)
still much better than the previous version

andrewliebenow replied:
"--base16" is one of the areas I found where "we" don't beat GNU. On my machine, base16 encoding is about 5-10% slower than GNU. It looks like it's more like 20% slower on your machine. I did make a note of that in the readme: "➖ Slower than GNU Core Utilities". Please let me know if you have any suggested changes here.

sylvestre commented on src/uu/basenc/BENCHMARKING.md:
```diff
@@ -0,0 +1,177 @@
+  Range (min … max):    2.649 s …  2.718 s    10 runs
+
+Relative speed comparison
+        2.67 ±  0.04  /usr/bin/basenc --base32hex --decode -- /dev/shm/one-random-gibibyte-base32hex-encoded 1>/dev/null
```
```
        3.27 ±  0.57  /usr/bin/basenc --base32hex --decode -- /dev/shm/one-random-gibibyte-base32hex-encoded 1>/dev/null
        1.00          ./target/release/coreutils basenc --base32hex --decode -- /dev/shm/one-random-gibibyte-base32hex-encoded 1>/dev/null
        1.04 ±  0.17  ./coreutils basenc --base32hex --decode -- /dev/shm/one-random-gibibyte-base32hex-encoded 1>/dev/null
```

getting something similar here :)

andrewliebenow replied:
Yep, looks like in some scenarios the new implementation isn't much faster. The main advantage here would just be the constant memory usage, and faster handling of "--ignore-garbage".

```
❯ hyperfine \
    --sort \
    command \
    -- \
    '/usr/bin/basenc --base32hex --decode --ignore-garbage -- /dev/shm/one-random-gibibyte-base32hex-encoded 1>/dev/null' \
    'coreutils basenc --base32hex --decode --ignore-garbage -- /dev/shm/one-random-gibibyte-base32hex-encoded 1>/dev/null' \
    './target/release/basenc --base32hex --decode --ignore-garbage -- /dev/shm/one-random-gibibyte-base32hex-encoded 1>/dev/null'

Benchmark 1: /usr/bin/basenc --base32hex --decode --ignore-garbage -- /dev/shm/one-random-gibibyte-base32hex-encoded 1>/dev/null
  Time (mean ± σ):     10.450 s ±  0.201 s    [User: 10.060 s, System: 0.327 s]
  Range (min … max):   10.242 s … 10.876 s    10 runs
 
Benchmark 2: coreutils basenc --base32hex --decode --ignore-garbage -- /dev/shm/one-random-gibibyte-base32hex-encoded 1>/dev/null
  Time (mean ± σ):     26.267 s ±  0.225 s    [User: 24.536 s, System: 1.566 s]
  Range (min … max):   25.986 s … 26.675 s    10 runs
 
Benchmark 3: ./target/release/basenc --base32hex --decode --ignore-garbage -- /dev/shm/one-random-gibibyte-base32hex-encoded 1>/dev/null
  Time (mean ± σ):      2.703 s ±  0.019 s    [User: 2.485 s, System: 0.201 s]
  Range (min … max):    2.664 s …  2.725 s    10 runs
 
Relative speed comparison
        3.87 ±  0.08  /usr/bin/basenc --base32hex --decode --ignore-garbage -- /dev/shm/one-random-gibibyte-base32hex-encoded 1>/dev/null
        9.72 ±  0.11  coreutils basenc --base32hex --decode --ignore-garbage -- /dev/shm/one-random-gibibyte-base32hex-encoded 1>/dev/null
        1.00          ./target/release/basenc --base32hex --decode --ignore-garbage -- /dev/shm/one-random-gibibyte-base32hex-encoded 1>/dev/null
```
